README.md
<!-- modrinth_exclude.start -->

![Create Track Map](https://raw.githubusercontent.com/gist/jenchanws/44291eb9435be4fc46be1b8661ae4406/raw/b545411b55a9b18bb42c58dcf22f909d2c404498/ctm.svg)

<!-- modrinth_exclude.end -->

# Create Track Map

![Mod version](https://img.shields.io/modrinth/v/gxoNIjg6)
![Minecraft versions](https://img.shields.io/badge/minecraft-1.18.2%20•%201.19.2-blue)
![Download count](https://img.shields.io/modrinth/dt/gxoNIjg6)

![Available for Fabric](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/522a26e2c91cbafab79ddbbc6d942f1720249a8c/fabric.svg)
[![Requires Create ≥0.5](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/522a26e2c91cbafab79ddbbc6d942f1720249a8c/create-0.5.svg)](https://modrinth.com/mod/create-fabric)
[![Requires Fabric API](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/1fc7a8dfba76460bedba7fa0b5516b917815e39c/fabric-api.svg)](https://modrinth.com/mod/fabric-api)
[![Requires Fabric Kotlin](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/1fc7a8dfba76460bedba7fa0b5516b917815e39c/fabric-kotlin.svg)](https://modrinth.com/mod/fabric-language-kotlin)

![Available for Forge](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/522a26e2c91cbafab79ddbbc6d942f1720249a8c/forge.svg)
[![Requires Create ≥0.5](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/522a26e2c91cbafab79ddbbc6d942f1720249a8c/create-0.5.svg)](https://modrinth.com/mod/create)
[![Requires Kotlin for Forge](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/522a26e2c91cbafab79ddbbc6d942f1720249a8c/forge-kotlin.svg)](https://modrinth.com/mod/kotlin-for-forge)

[![Available on GitHub](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/0dbefc2fcbec362d14f1689acb807183ceffdbe1/github.svg)](https://github.com/jenchanws/create-track-map)
[![Available on Modrinth](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/0dbefc2fcbec362d14f1689acb807183ceffdbe1/modrinth.svg)](https://modrinth.com/mod/create-track-map)
[![Find me on Discord](https://raw.githubusercontent.com/gist/jenchanws/842eee8428e1e0aec20de4594878156a/raw/0dbefc2fcbec362d14f1689acb807183ceffdbe1/discord.svg)](https://smp.littlechasiu.com/discord)

Create Track Map (CTM) is a multi-loader mod that displays a track map of Create trains in your world, including all tracks, signals, stations, and trains. The signals and trains are updated in (practically) real time.

![Example track map](https://cdn.modrinth.com/data/gxoNIjg6/images/8aa58af4ca9cc459a84ce492770a92e358cd2714.gif)

# Usage

CTM is available for [download on Modrinth](https://modrinth.com/mod/create-track-map). **Please make sure you download the correct version for your loader, Minecraft version, and Create version.** Below is the compatibility chart for Create versions:

| | CTM ≤1.3 | CTM 1.4 | CTM ≥1.5 |
|:---|:--:|:--:|:--:|
| **Create 0.5.0** | ✅ | ✅ | ❌ |
| **Create 0.5.1** | ❌ | ✅ | ✅ |

## Installation
Simply place the .jar file in your mods folder (or install with a modloader) like any other mod! CTM is intended to be a server side mod, but can also run in single-player worlds and LAN servers. It runs a web server on port `3876` by default, but [this can be configured](#configuration). You must forward this port if you wish to see the map. *[(How do I do this?)](https://www.google.com/search?q=how+to+I+port+forward)* 

The map itself is visible at the root (by default `http://localhost:3876/`). If using a proxy to serve the map, make sure to configure it to let Server-Sent Events through.

### *Singleplayer*
The map is hosted on your computer so while Minecraft is open, enter `http://localhost:3876/` into your browser.

### *Multiplayer*

 **A) Local (LAN) server**\
The map is still hosted on your computer so while the game is open, enter `http://localhost:3876/` into your browser.\
Those in your household can see the map by entering `http://[your computer's IPV4 address]:3876/` into their browser.\
*[How do I see my computer's IPV4 address?](https://www.google.com/search?q=how+do+i+see+my+local+ipv4+address)*

**B) The server is on my PC**\
The map is still hosted on your computer so while the server is open,  enter `http://localhost:3876/` into your browser.\
Those in your household can see the map by entering `http://[your computer's IPV4 address]:3876/` into their browser.\
Those outside your household can see the map by entering `http://[your public IPV4 address]:3876/` into their browser.\
*[How do I see my public IPV4 address?](https://www.google.com/search?q=what+is+my+public+ipv4+address)*

**C) The server is hosted with a hosting service**\
The map is hosted with the server on that service. You can see the map at `[your server address]:3876/` while the server is up.\
**NOTE: YOUR SERVICE MUST SUPPORT EXTRA PORT FORWARDING.** You can check with support/the docs for the service or just good old Google. If the service doesn't support this, you are out of luck. You can switch services or try hosting your own.

*Just so it's clear: Do not include the brackets around your IPV4 address! For example: `192.168.0.0:3876`*

## READ THIS BEFORE ASKING FOR HELP!

To use this mod, **you must be able to open a second port**, in addition to your Minecraft server, which CTM will run its web server on. If you are using Aternos, this is not possible. If you are using a shared hosting provider, please contact your host for assistance in opening a port. No support will be provided to open a port or to diagnose issues which are not caused by a bug in the mod.

# Advanced Usage

### CTM API
CTM provides the following API, also accessible through the same port:

- `/api/network`, `/api/network.rt`: List of all track pieces and train stations
- `/api/signals`, `/api/signals.rt`: List of all train signals, including their states
  (green, yellow, red)
- `/api/blocks`, `/api/blocks.rt`: List of all signal control blocks, and whether they are occupied or reserved by a train
- `/api/trains`, `/api/trains.rt`: List of all assembled trains, including their names and
  positions
- `/api/style.css`: CSS style sheet generated from configured colors and fonts
- `/api/config.json`: Map configuration

The `.rt` versions update in real time with Server-Sent Events (SSE).

### Configuration

CTM's config options can be found at `create-track-map.json` in your server's config directory. It is automatically created at startup if it doesn't exist.

The following options are available:

```js
{
  // Whether to actually start the watcher and the server.
  "enable": false,

  // How long to wait between track data updates.
  "watch_interval_seconds": 0.5,
  // The port the internal web server listens on.
  "server_port": 3876,

  "map_style": {
    // Font to use for the map's UI. Must be a valid CSS font stack.
    "font": "ui-monospace, \"JetBrains Mono\", monospace",
    // Colors for individual components of the map. Must be valid CSS colors.
    // Any CSS color format will work, such as named colors and rgb().
    "colors": {
      "background": "#888",
      "track": {
        "occupied": "red",
        "reserved": "pink",
        "free": "white"
      },
      "signal": {
        "green": "#71db51",
        "yellow": "#ffd15c",
        "red": "#ff5f5c",
        "outline": "black"
      },
      "portal": {
        "primary": "purple",
        "outline": "white"
      },
      "station": {
        "primary": "white",
        "outline": "black"
      },
      "train": "cyan",
      "lead_car": "darkturquoise"
    }
  },

  "map_view": {
    "initial_dimension": "minecraft:overworld",
    "initial_position": { "x": 0, "z": 0 },

    // Zoom levels must be integers, but may be negative.
    // Each zoom level is twice as big as the previous.
    // 0 is a decent minimum but may be impractical for large networks.
    // 3 is the sensible default for viewing double-tracked networks.
    "initial_zoom": 3,
    "min_zoom": 0,
    "max_zoom": 4,

    // Whether a zoom control should be visible on the screen.
    "zoom_controls": true,

    // Which side of the track a signal should be displayed on.
    // Valid values are LEFT or RIGHT (all uppercase).
    "signals_on": "RIGHT"
  },

  "dimensions": {
    // Dimension names must be namespaced.
    "minecraft:overworld": {
      // Label that shows up in the layer switcher.
      "label": "Overworld"
    },
    "minecraft:the_nether": {
      "label": "Nether"
    },
    "minecraft:the_end": {
      "label": "End"
    }
  },

  "layers": {
    "tracks": {
      "label": "Tracks",
      "min_zoom": 0,
      "max_zoom": 4
    },
    "blocks": {
      "label": "Track Occupancy",
      "min_zoom": 0,
      "max_zoom": 4
    },
    "signals": {
      "label": "Signals",
      "min_zoom": 0,
      "max_zoom": 4
    },
    "portals": {
      "label": "Portals",
      "min_zoom": 0,
      "max_zoom": 4
    },
    "stations": {
      "label": "Stations",
      "min_zoom": 0,
      "max_zoom": 4
    },
    "trains": {
      "label": "Trains",
      "min_zoom": 0,
      "max_zoom": 4
    }
  }
}
```

Reload the config without restarting the server by running `/ctm reload` (operator permissions required).


CHANGELOG.md
- Added support for Minecraft 1.20.1.


src/main/kotlin/littlechasiu/ctm/model/Config.kt
package littlechasiu.ctm.model

import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
@OptIn(ExperimentalSerializationApi::class)
data class SignalColors(
  @EncodeDefault
  val green: String = "#71db51",
  @EncodeDefault
  val yellow: String = "#ffd15c",
  @EncodeDefault
  val red: String = "#ff5f5c",
  @EncodeDefault
  val outline: String = "black",
)

@Serializable
@OptIn(ExperimentalSerializationApi::class)
data class TrackColors(
  @EncodeDefault
  val occupied: String = "red",
  @EncodeDefault
  val reserved: String = "pink",
  @EncodeDefault
  val free: String = "white",
)

@Serializable
data class IconColors(
  val primary: String,
  val outline: String,
)

@Serializable
@OptIn(ExperimentalSerializationApi::class)
data class Colors(
  @EncodeDefault
  val background: String = "#888",
  @EncodeDefault
  val track: TrackColors = TrackColors(),
  @EncodeDefault
  val signal: SignalColors = SignalColors(),
  @EncodeDefault
  val portal: IconColors = IconColors(primary = "purple", outline = "white"),
  @EncodeDefault
  val station: IconColors = IconColors(primary = "white", outline = "black"),
  @EncodeDefault
  val train: String = "cyan",
  @SerialName("lead_car")
  @EncodeDefault
  val leadCar: String = "plum",
)

@Serializable
@OptIn(ExperimentalSerializationApi::class)
data class MapStyle(
  @EncodeDefault
  val font: String = "ui-monospace, \"JetBrains Mono\", monospace",
  @EncodeDefault
  val colors: Colors = Colors(),
)

@Serializable
data class Coordinates(
  val x: Int,
  val z: Int,
)

@Serializable
enum class TrackSide {
  LEFT,
  RIGHT,
}

@Serializable
@OptIn(ExperimentalSerializationApi::class)
data class MapView(
  @SerialName("initial_dimension")
  @EncodeDefault
  val initialDimension: String = "minecraft:overworld",
  @SerialName("initial_position")
  @EncodeDefault
  val initialPosition: Coordinates = Coordinates(0, 0),
  @SerialName("initial_zoom")
  @EncodeDefault
  val initialZoom: Int = 3,

  @SerialName("min_zoom")
  @EncodeDefault
  val minZoom: Int = 0,
  @SerialName("max_zoom")
  @EncodeDefault
  val maxZoom: Int = 4,

  @SerialName("zoom_controls")
  @EncodeDefault
  val zoomControls: Boolean = false,

  @SerialName("signals_on")
  @EncodeDefault
  val signalSide: TrackSide = TrackSide.RIGHT,
)

@Serializable
@OptIn(ExperimentalSerializationApi::class)
data class LayerConfig(
  val label: String,
  @SerialName("min_zoom")
  @EncodeDefault
  val minZoom: Int = 0,
  @SerialName("max_zoom")
  @EncodeDefault
  val maxZoom: Int = 4,
)

@Serializable
data class DimensionConfig(
  val label: String,
)

@Serializable
data class MapConfig(
  val view: MapView,
  val dimensions: Map<String, DimensionConfig>,
  val layers: Map<String, LayerConfig>,
)

@Serializable
data class Config @OptIn(ExperimentalSerializationApi::class) constructor(
  @EncodeDefault
  val enable: Boolean = true,

  @SerialName("watch_interval_seconds")
  @EncodeDefault
  val watchIntervalSeconds: Double = 0.5,
  @SerialName("server_port")
  @EncodeDefault
  val serverPort: Int = 3876,

  @SerialName("map_style")
  @EncodeDefault
  val mapStyle: MapStyle = MapStyle(),
  @SerialName("map_view")
  @EncodeDefault
  val mapView: MapView = MapView(),
  @EncodeDefault
  val dimensions: Map<String, DimensionConfig> = mapOf(
    "minecraft:overworld" to DimensionConfig(label = "Overworld"),
    "minecraft:the_nether" to DimensionConfig(label = "Nether"),
    "minecraft:the_end" to DimensionConfig(label = "End"),
  ),
  @EncodeDefault
  val layers: Map<String, LayerConfig> = mapOf(
    "tracks" to LayerConfig(label = "Tracks"),
    "blocks" to LayerConfig(label = "Track Occupancy"),
    "signals" to LayerConfig(label = "Signals"),
    "portals" to LayerConfig(label = "Portals"),
    "stations" to LayerConfig(label = "Stations"),
    "trains" to LayerConfig(label = "Trains"),
  ),
)
